### PR TITLE
Dump and load endpoints

### PR DESCRIPTION
**Pull request by mikiw**
## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/starknet/dump.rs [hunk 1]
```diff
@@ -176,6 +191,9 @@ impl Starknet {
                     }
 
                     Ok(transactions)
+                } else if custom_path.is_some() {
+                    // in case of load from http endpoint return error
```
Comment (ivpavici) [inline crates/starknet/src/starknet/dump.rs:195]: this same comment is also present above so maybe redundant?

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs [hunk 2]
```diff
@@ -1,19 +1,38 @@
...
 pub(crate) async fn load(
-    Json(_path): Json<Path>,
-    Extension(_state): Extension<HttpApiHandler>,
-) -> HttpApiResult<()> {
-    Err(HttpApiError::PathNotFound)
+    Json(path): Json<Path>,
+    Extension(state): Extension<HttpApiHandler>,
+) -> HttpApiResult<Json<DumpLoadResponse>> {
+    let file_path = std::path::Path::new(&path.path);
+    if path.path.is_empty() || !file_path.exists() {
```
Comment (ivpavici) [inline crates/starknet-server/src/api/http/endpoints/dump_load.rs:27]: is `path.path` usual, can the name be different? just looks strange to me as a non-Rust person

File: crates/starknet/src/starknet/dump.rs (hunk 1 above)
Comment (mikiw) [reply crates/starknet/src/starknet/dump.rs:195]: yes not it is, thanks :D

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs [hunk 3]
```diff
@@ -1,19 +1,38 @@
...
 pub(crate) async fn dump(
-    Json(_path): Json<Path>,
-    Extension(_state): Extension<HttpApiHandler>,
-) -> HttpApiResult<()> {
-    Err(HttpApiError::GeneralError)
+    Json(path): Json<Path>,
+    Extension(state): Extension<HttpApiHandler>,
+) -> HttpApiResult<Json<DumpLoadResponse>> {
+    if path.path.is_empty() {
+        return Err(HttpApiError::PathNotFound);
```
Comment (FabijanC) [inline crates/starknet-server/src/api/http/endpoints/dump_load.rs:12]: In Devnet-py we allowed path to be empty if it was specified with `--dump-path`:
https://github.com/0xSpaceShard/starknet-devnet/blob/ef036e26fac86ff197e8fecd0550f62776d511d2/starknet_devnet/blueprints/base.py#L91

Did you consult the Python implementation (also for tests)?

EDIT: I am now coming back to this comment confused after seeing the defaulting done with:
```rust
if custom_path.is_some() { &custom_path } else { &self.config.dump_path }
```

File: crates/starknet/src/starknet/dump.rs [hunk 4]
```diff
@@ -129,9 +129,14 @@ impl Starknet {
...
 
-    /// save starknet transactions to file
     pub fn dump_transactions(&self) -> DevnetResult<()> {
-        match &self.config.dump_path {
+        self.dump_transactions_custom_path(None)
+    }
+
+    /// save starknet transactions to file
+    pub fn dump_transactions_custom_path(&self, custom_path: Option<String>) -> DevnetResult<()> {
+        let dump_path = if custom_path.is_some() { &custom_path } else { &self.config.dump_path };
```
Comment (FabijanC) [inline crates/starknet/src/starknet/dump.rs:138]: Can't `unwrap_or` be used here?

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 5]
```diff
@@ -217,4 +220,54 @@ mod dump_and_load_tests {
...
     }
+
+    #[tokio::test]
+    async fn dump_load_endpoint() {
+        let dump_file_name = "dump_endpoint";
+        let devnet_dump = BackgroundDevnet::spawn_with_additional_args(&[
+            "--dump-path",
+            dump_file_name,
+            "--dump-on",
+            "exit",
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:231]: Why is it started with `--dump-on`?

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs]: yes you are right, fixed in c24c7dd6e52159f12b71fb47787dfedef2cfd37a

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs [hunk 6]
```diff
@@ -1,19 +1,38 @@
...
+    }
+    let starknet = state.api.starknet.write().await;
+    starknet
+        .dump_transactions_custom_path(Some(path.path.clone()))
+        .map_err(|_| HttpApiError::GeneralError)?;
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs]: I added new error types bcc7ef730f8b032ab2abeeb14bc21ae6d4f6de08

File: crates/starknet-server/src/api/http/models.rs [hunk 7]
```diff
@@ -97,3 +97,8 @@ pub(crate) struct ForkStatus {
     url: String,
     block: u128,
 }
+
+#[derive(Serialize)]
+pub(crate) struct DumpLoadResponse {
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/http/models.rs]: Fixed in b755bdd48d46d2ab5d69061eb53d5727aa310a18

File: crates/starknet/src/starknet/dump.rs [hunk 8]
```diff
@@ -145,23 +150,33 @@ impl Starknet {
...
+        match dump_path {
             Some(path) => {
                 let file_path = Path::new(path);
 
-                // load only if the file exists, if dump_path is set but the file doesn't exist it
-                // can mean that it's first run with dump_path parameter set to dump, in that case
-                // return empty vector
+                // load only if the file exists, if config.dump_path is set but the file doesn't
+                // exist it means that it's first execution and in that case return an empty vector,
+                // in case of load from HTTP endpoint return FileNotFound error
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet/src/starknet/dump.rs:179]: Fixed in 4a89d5352d260962337b4cd90567453d3e2f5455

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 5 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:231]: right, it can just be `BackgroundDevnet::spawn()`

Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in 5218df73f399e95a2cf17d76c6f6ce3ca7003fe6

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 9]
```diff
@@ -217,4 +220,54 @@ mod dump_and_load_tests {
...
+        let load_body = Body::from(
+            json!({
+                "path": dump_file_name
+            })
+            .to_string(),
+        );
+        devnet_load.post_json("/load".into(), load_body).await.unwrap();
+
+        let loaded_transaction =
+            devnet_load.json_rpc_client.get_transaction_by_hash(mint_tx_hash).await.unwrap();
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:265]: Fixed in 99ad79d2ef3414addad735884b58d807e226065b

File: crates/starknet/src/starknet/dump.rs (hunk 4 above)
Comment (mikiw) [reply crates/starknet/src/starknet/dump.rs:138]: I don't this so since dump_path is Option<String>

File: crates/starknet/src/starknet/dump.rs (hunk 1 above)
Comment (mikiw) [reply crates/starknet/src/starknet/dump.rs]: removed in c24c7dd6e52159f12b71fb47787dfedef2cfd37a

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs (hunk 2 above)
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs:28]: I think it's unavoidable since it's the local variable name of `struct Path` that contains `path` property that it's mapped from JSON. I could only change the local variable name to be `file` so it would be `file.path` but it's inappropriate and misleading in this case since there is no `File` in this context and it's just information about the path.

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs (hunk 3 above)
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs]: I was briefly checking Python implementation at the beginning of that task but to be honest, this re-execution approach is a completely different way of doing this dump and load.

Fixed in 4d89670c7d9c891765ce73533e82e83eb0908c54

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 10]
```diff
@@ -217,4 +220,54 @@ mod dump_and_load_tests {
 
         remove_file(dump_file_name);
     }
+
+    #[tokio::test]
+    async fn dump_load_endpoint() {
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: I'll add some edge cases

[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in 1e74d364a715411c6f86856347d5aefd728d4b6b

File: README.md [hunk 11]
```diff
@@ -139,6 +139,11 @@ cargo run -- --dump-on exit --dump-path <PATH>
 cargo run -- --dump-on transaction --dump-path <PATH>
 ```
 
+Dumping on request (replace <HOST>, <PORT> and <PATH> with your own):
```
Comment (FabijanC) [inline README.md:142]: Look at other documented cases (Dumping on exit, Dumping after each transaction), they each have a dash in the beginning of the line.

File: README.md [hunk 12]
```diff
@@ -153,6 +158,11 @@ Currently, dumping produces a list of received transactions that is stored on di
 Conversely, loading is implemented as the re-execution of transactions from a dump.
 This means that timestamps of `StarknetBlock` will be different.
 
+Loading on request:
```
[the remark this answers is not in the document]
Comment (mikiw) [reply README.md]: Fixed in f0e89403cd85fcc72ccff8087642f628806fe339

File: README.md (hunk 11 above)
Comment (mikiw) [reply README.md]: Fixed in f0e89403cd85fcc72ccff8087642f628806fe339

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 13]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+        }
+
+        #[cfg(unix)]
+        {
+            let mut kill = Command::new("kill")
+                .args(["-s", "SIGINT", &devnet_dump.process.id().to_string()])
+                .spawn()
+                .unwrap();
+            kill.wait().unwrap();
+        }
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:281]: Can't L264-L281 be extracted into a separate funcitonand used in all places requiring process killing? I think there is another place with this exact same code.

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 14]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+    async fn dump_endpoint_fail_with_wrong_file_name() {
+        let devnet_dump = BackgroundDevnet::spawn().await.expect("Could not start Devnet");
+        let dump_body = Body::from(
+            json!({
+                "path": "///"
+            })
+            .to_string(),
+        );
+        let result = devnet_dump.post_json("/dump".into(), dump_body).await.unwrap();
+        assert_eq!(result.status(), 500);
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:328]: Why is it 500 here? It means we messed something up. Perhaps we need to improve handling of invalid paths.

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 15]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+        let dump_file_name = "dump_endpoint";
+        let dump_file_name_custom_path = "dump_endpoint_custom_path";
+        let devnet_dump =
+            BackgroundDevnet::spawn_with_additional_args(&["--dump-path", dump_file_name])
+                .await
+                .expect("Could not start Devnet");
+        let mint_tx_hash = devnet_dump.mint(DUMMY_ADDRESS, DUMMY_AMOUNT).await;
+        let dump_body = Body::from(
+            json!({
+                "path": ""
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:356]: Ok, so it works if path is an empty string. But what if it's omitted - didn't we allow that in devnet-py?

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 16]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+            .to_string(),
+        );
+        let result = devnet_load.post_json("/load".into(), load_body).await.unwrap();
+        assert_eq!(result.status(), 400);
+    }
+
+    #[tokio::test]
+    async fn dump_load_endpoints_transaction_and_state_after_load_is_valid() {
+        let dump_file_name = "dump_endpoint";
+        let dump_file_name_custom_path = "dump_endpoint_custom_path";
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:348]: No need to define the variable so early.

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 17]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+                "path": load_file_name
+            })
+            .to_string(),
+        );
+        let result = devnet_load.post_json("/load".into(), load_body).await.unwrap();
+        assert_eq!(result.status(), 400);
+    }
+
+    #[tokio::test]
+    async fn dump_load_endpoints_transaction_and_state_after_load_is_valid() {
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:346]: This test could use some empty lines and two or three simple `// explanation comments`

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 18]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+            remove_file(dump_file_name);
+        } else {
+            panic!("Could not find dump file {}", dump_file_name);
+        }
+        if Path::new(dump_file_name_custom_path).exists() {
+            remove_file(dump_file_name_custom_path);
+        } else {
+            panic!("Could not find dump file {}", dump_file_name_custom_path);
+        }
+    }
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:416]: Why another round of existence assertion? Wasn't it already checked at this point?

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:396]: let's leave it as it is, seems that it's not implemented:

```
pub(crate) async fn get_account_balance(
    Query(_contract_address): Query<ContractAddress>,
) -> HttpApiResult<Json<Balance>> {
    Err(HttpApiError::GeneralError)
}
```

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Great testing improvement. Will resolve this but I'm about to leave more comments in the new test.

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 19]
```diff
@@ -217,4 +247,171 @@ mod dump_and_load_tests {
...
+                .await
+                .expect("Could not start Devnet");
+        let mint_tx_hash = devnet_dump.mint(DUMMY_ADDRESS, DUMMY_AMOUNT).await;
+        let dump_body = Body::from(
+            json!({
+                "path": ""
+            })
+            .to_string(),
+        );
+        devnet_dump.post_json("/dump".into(), dump_body).await.unwrap();
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_dump_and_load.rs:360]: I don't understand, what is the purpose of this second dumping in this test? We aren't loading it anywhere, right? Should it be a separate case? Again, I might be wrong because I don't understand the intended purpose.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 15 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:356]: If it's omitted it will not pass JSON validation, but I'll check this case.

Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs:356]: I think we should either accept the omitted property, or not accept empty strings at all.

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 16 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:348]: I wanted to have them near, so I could be sure that they were different. Also, this will change during https://github.com/0xSpaceShard/starknet-devnet-rs/issues/205

Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs:348]: Ok, makes sense. You can resolve this conversation if you want.

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 19 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:360]: The first dump should take the default path which is "dump_endpoint" and we asserting that. Secondly, we check the second dump with a custom path "dump_endpoint_custom_path".

Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:360]: I'll add comments with that

Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:362]: Added comments here 63e664ee8429af41b9ca690d52682f0f2dae2f31

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 18 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in 63e664ee8429af41b9ca690d52682f0f2dae2f31

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 17 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:346]: Fixed in 63e664ee8429af41b9ca690d52682f0f2dae2f31

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 15 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in 04ca6f2e8a6b18388572e86e6adddf68f72a3f5b and 60bf67a4a136d6709707914f0aa5c5ab7899893e

Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Now if json body is empty or path is "" we will use default path.

Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: "not accept empty strings at all" Why is that? I think that an empty path is straightforward with the intention of using a default path.

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 13 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in 04ca6f2e8a6b18388572e86e6adddf68f72a3f5b

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs [hunk 20]
```diff
@@ -1,19 +1,39 @@
...
-    Json(_path): Json<Path>,
-    Extension(_state): Extension<HttpApiHandler>,
-) -> HttpApiResult<()> {
-    Err(HttpApiError::GeneralError)
+    Json(path): Json<Path>,
+    Extension(state): Extension<HttpApiHandler>,
+) -> HttpApiResult<Json<DumpResponse>> {
+    let starknet = state.api.starknet.write().await;
+    if path.path.is_empty() {
+        starknet.dump_transactions().map_err(|_| HttpApiError::DumpError)?;
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs]: Fixed in e9bf7062aa1586902d158bb8a316c6b867c6c75f

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 14 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Fixed in e9bf7062aa1586902d158bb8a316c6b867c6c75f

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 15 above)
Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs]: Straightforward would mean explicit, but here it's implicit (if empty string - use default). A user passing an empty string could also mean that they accidentally parsed something wrong. But I'm ok with keeping this implication if that's how it worked in devnet-py.

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs [hunk 21]
```diff
@@ -1,19 +1,54 @@
...
+    Extension(state): Extension<HttpApiHandler>,
+) -> HttpApiResult<Json<DumpResponse>> {
+    let starknet = state.api.starknet.write().await;
+    match path.path {
+        None => {
+            // path not present
+            starknet
+                .dump_transactions()
+                .map_err(|err| HttpApiError::DumpError { msg: err.to_string() })?;
+            Ok(Json(DumpResponse { path: "".to_string() }))
```
Comment (FabijanC) [inline crates/starknet-server/src/api/http/endpoints/dump_load.rs:18]: Why are we returning an empty path to the user? I'm almost sure that's not what we did in devnet-py?

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 19 above)
Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs:345]: But why are there two dumps and only one load? Also, is there a typo in this comment (is -> if):
```
load and re-execute from "dump_endpoint" file and check is transaction
```

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 15 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs]: https://github.com/0xSpaceShard/starknet-devnet/blob/e477aa1bbe2348ba92af2a69c32d2eef2579d863/starknet_devnet/blueprints/base.py#L86

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 19 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_dump_and_load.rs:345]: I'm testing functionality if /dump endpoint with an empty JSON body works and the default path is taken in that case. Later I'm testing if I can still set the custom path and dump to the custom path and the path will be different (not default). Files are the same so I could load any of them later.

Typo fixed in 0e1bd79aa51cd93ea82d372e18a1256cae10c6ee

File: crates/starknet-server/src/api/http/endpoints/dump_load.rs (hunk 21 above)
Comment (mikiw) [reply crates/starknet-server/src/api/http/endpoints/dump_load.rs]: right, in the old devnet we just return `Response(status=200)` so I fixed that to be `HttpApiResult<()>` in 2565cb00669394c78f6097a9bf0fd789f9b0bc68

File: crates/starknet-server/tests/test_dump_and_load.rs (hunk 15 above)
Comment (FabijanC) [reply crates/starknet-server/tests/test_dump_and_load.rs]: It's ok. In devnet-py the defaulting was done if either it was an empty string or if it was `None` (i.e. undefined). Resolving this.

**Review by FabijanC — APPROVED** [review]
